microdnf/commands/upgrade/upgrade.cpp

```diff
@@ -81,24 +83,28 @@ void CmdUpgrade::run(Context & ctx) {
 
     std::cout << std::endl;
 
-    libdnf::rpm::PackageSet result_pset(&solv_sack);
-    libdnf::rpm::SolvQuery full_solv_query(&solv_sack);
+    libdnf::Goal goal(&ctx.base);
     for (auto & pattern : *patterns_to_upgrade_options) {
-        libdnf::rpm::SolvQuery solv_query(full_solv_query);
-        solv_query.resolve_pkg_spec(dynamic_cast<libdnf::OptionString *>(pattern.get())->get_value(), true, true, true, true, true, {});
-        result_pset |= solv_query.get_package_set();
+        goal.add_rpm_upgrade(dynamic_cast<libdnf::OptionString *>(pattern.get())->get_value(), {});
     }
+    goal.resolve();
 
-    download_packages(result_pset, nullptr);
+    if (!print_goal(goal)) {
+        return;
+    }
 
-    std::vector<std::unique_ptr<RpmTransactionItem>> transaction_items;
-    auto ts = libdnf::rpm::Transaction(ctx.base);
-    for (auto package : result_pset) {
-        auto item = std::make_unique<RpmTransactionItem>(package, RpmTransactionItem::Actions::UPGRADE);
-        auto item_ptr = item.get();
-        transaction_items.push_back(std::move(item));
-        ts.upgrade(*item_ptr);
+    std::cout << "Is this ok [y/N]: ";
+    std::string answer;
+    std::getline(std::cin, answer);
+    if (answer.size() != 1 || (answer[0] != 'y' && answer[0] != 'Y')) {
+        std::cout << "Operation aborted." << std::endl;
+        return;
     }
+
+    std::vector<std::unique_ptr<RpmTransactionItem>> transaction_items;
+    libdnf::rpm::Transaction ts(ctx.base);
+    prepare_transaction(goal, ts, transaction_items);
```

Comment:
It looks like that download rpm packages step disappeared?

Reply:
Packages are downloaded automaticaly in `prepare_transaction()` now. I will move the code to another `download_packages()` function.